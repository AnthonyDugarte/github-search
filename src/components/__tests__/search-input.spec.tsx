import React from "react"
import { render, fireEvent, screen } from "@testing-library/react"
import faker from "faker"

import SearchInput from "../search-input"

const githubSearchInputTestId = "github-search-input"

describe("Search Input component", () => {
  // Common vars to be used between tests
  let mockOnSearch, renderedSearchInput

  // Before each test we populate our rendered and mocked elements to be used at testing
  beforeEach(() => {
    mockOnSearch = jest.fn<void, [string]>()
    renderedSearchInput = render(<SearchInput onSearch={mockOnSearch} />)
  })

  test("Matches snapshot", () => {
    expect(renderedSearchInput.asFragment()).toMatchSnapshot()
  })

  test("Correctly debounce searching", async () => {
    await testDebouncedInput(faker.name.firstName(), mockOnSearch, 0)
    await testDebouncedInput(faker.name.firstName(), mockOnSearch, 1)
  })
})

/**
 * Simulate user input
 * @param value string to input
 */
function dispatchInputEntry(value: string): void {
  fireEvent.input(screen.getByTestId(githubSearchInputTestId), {
    target: { value: value },
  })
}

async function testDebouncedInput(
  search: string,
  onSearchMock: any,
  callIdx: number
): Promise<void> {
  // Dispatch simulation of input
  dispatchInputEntry(search)

  // User should see what he is typing
  expect(screen.getByTestId(githubSearchInputTestId)).toHaveValue(search)

  // No searching has been made yet, waiting for user to stop typing
  expect(onSearchMock.mock.calls[callIdx]).toBeUndefined()
  await new Promise(resolve => setTimeout(resolve, 100))
  expect(onSearchMock.mock.calls[callIdx]).toBeUndefined()

  // "Wait" for user to stop typing
  await new Promise(resolve => setTimeout(resolve, 500))

  // A search should have been made
  expect(onSearchMock.mock.calls[callIdx]).not.toBeUndefined()
  expect(onSearchMock.mock.calls[callIdx][0]).toBe(search)
}
