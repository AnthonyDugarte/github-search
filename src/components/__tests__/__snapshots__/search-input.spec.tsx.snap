// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Input component Matches snapshot 1`] = `
<DocumentFragment>
  <input
    class="border focus:outline-none focus:shadow-outline rounder border-gray-300 rounded-lg py-2 px-4 block appearance-none leading-normal w-full lg:w-2/3"
    data-testid="github-search-input"
    placeholder="Search for someone..."
    type="text"
    value=""
  />
</DocumentFragment>
`;
